app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/AddPictureActivity.kt

```diff
@@ -0,0 +1,126 @@
+import androidx.camera.core.CameraSelector
+import androidx.camera.core.Preview
+import androidx.camera.lifecycle.ProcessCameraProvider
+import androidx.camera.view.PreviewView
+import androidx.core.content.ContextCompat
+import androidx.lifecycle.LifecycleOwner
+import com.google.common.util.concurrent.ListenableFuture
+
+/*
+Activity where an administrator can take a picture to add it to a selected data set
+ */
+class AddPictureActivity : AppCompatActivity() {
+
+    private lateinit var cameraProviderFuture: ListenableFuture<ProcessCameraProvider>
+    private lateinit var requestPermissionLauncher: ActivityResultLauncher<String>
+    private lateinit var openLabelSelectionLayoutListener: View.OnClickListener
+
+    private val dummyLabels = arrayOf("label1", "label2", "label3", "label4", "label5", "label6", "label7", "label8", "label9", "label10")
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setupRequestPermissionLauncher()
+        setupOpenLabelSelectionLayoutListener()
+
+        if (cameraIsAvailable()) {
+            when {
+                ContextCompat.checkSelfPermission(
+                        this,
+                        Manifest.permission.CAMERA
+                ) == PackageManager.PERMISSION_GRANTED -> {
+                    setupCameraPreviewView()
+                }
+                shouldShowRequestPermissionRationale(Manifest.permission.CAMERA) -> {
+                    permissionNeededDialog()
+                }
+                else -> {
+                    requestPermissionLauncher.launch(Manifest.permission.CAMERA)
+                }
+            }
+        } else {
+            permissionNeededDialog()
+        }
+    }
+
+    private fun cameraIsAvailable(): Boolean {
+        return applicationContext.packageManager.hasSystemFeature(PackageManager.FEATURE_CAMERA)
+    }
+
+    private fun setupRequestPermissionLauncher() {
+        requestPermissionLauncher =
+                registerForActivityResult(ActivityResultContracts.RequestPermission()) { isGranted: Boolean ->
+                    if (isGranted) {
+                        setupCameraPreviewView()
+                        Toast.makeText(applicationContext, getString(R.string.AddPicture_permissionGrantedToast), Toast.LENGTH_SHORT).show()
+                    } else {
+                        permissionNeededDialog()
+                    }
+                }
+    }
+
+    /*
+    Needs to be set on a button
+     */
+    private fun setupOpenLabelSelectionLayoutListener() {
+        openLabelSelectionLayoutListener = View.OnClickListener { it: View ->
+            val builder = AlertDialog.Builder(this)
+
+            builder.apply {
+                setTitle(getString(R.string.AddPicture_labelSelectionDialogTitle))
+                setItems(dummyLabels) { dialog, label_index ->
+                    (it as Button).text = dummyLabels[label_index]
```

Comment:
This could be excess of caution, but I'm not very confortable with that down-cast. The alternative would be to go do `findViewById`, which is checked at compile time.

```suggestion
                    findViewById<Button>(R.id.selectLabelButton).text = dummyLabels[label_index]
```

Reply:
I agree and modified it as suggested.